### ISSUE TITLE
Problem with mock when using abstract class

### ISSUE THREAD

**Issue by simaoneves** [icy-arctic-fox/spectator#42]
When using an abstract class in the following example, i get an error that the method `register_hook` is not being implemented.

```crystal
require "./spec_helper"

abstract class SdkInterface
  abstract def register_hook(name, &block)
end

class Example
  def initialize(@sdk : Sdk)
    
  end

  def configure
    @sdk.register_hook("name") do
      nil
    end
  end
end

class Sdk < SdkInterface
  def initialize
  end

  def register_hook(name, &block)
    nil
  end
end

Spectator.describe Example do
  mock Sdk do
    stub register_hook(name, &block)
  end

  describe "#configure" do
    it "registers a block on configure" do
      sdk = Sdk.new
      example_class = Example.new(sdk)
      allow(sdk).to receive(register_hook())

      example_class.configure

      expect(sdk).to have_received(register_hook()).with("name")
    end
  end
end

```

If i change the `Sdk` class to not inherit from `SdkInterface` it works as expected.

**Comment by icy-arctic-fox** [comment]
Appears to be an issue with stubbing a method that takes a block. Removing all usages of `&block` and the block in `Example#configure` cause the spec to compile and pass. It looks like the injected stub code is omitting the block (probably never implemented). Since a method that takes a block is technically has a different signature, this explains the "not implemented" error.

I'm looking into the issue.

**Comment by icy-arctic-fox** [comment]
This should be addressed on `master`. Can you try that with your code to ensure it works as expected?

**Comment by simaoneves** [comment]
After updating to master, the example works great 😄 Many thanks!

I got into other problems though..
<img width="805" alt="Screenshot 2022-02-22 at 22 57 35" src="https://user-images.githubusercontent.com/1037529/155233658-e6d7fb30-ded1-44a9-a305-08fa8b4d382b.png">

with the following example:
```crystal
require "./spec_helper"

abstract class SdkInterface
  abstract def register_hook(name, &block)
end

class Example
  def initialize(@sdk : Sdk)
  end

  def configure
    @sdk.register_hook("name") do

    end
  end
end

class Sdk < SdkInterface
  def initialize
  end

  def register_hook(name, &block)
    other_method(&block)
    nil
  end

  def other_method(block)
  end
end

Spectator.describe Example do
  mock Sdk do
    stub register_hook(name, &block)
  end

  describe "#configure" do
    it "registers a block on configure" do
      sdk = Sdk.new
      example_class = Example.new(sdk)
      allow(sdk).to receive(register_hook())

      example_class.configure

      expect(sdk).to have_received(register_hook()).with("name")
    end
  end
end

```

Should i open another issue?

**Comment by icy-arctic-fox** [comment]
No need for another issue. I'll try to get this fixed soon.

**Comment by icy-arctic-fox** [comment]
Providing an update here since it's been a while.

I have decided enough is enough and I've invested effort into redoing the entire mock and stub system. Instead of trying to fix this issue with the current system, it will be fixed in the new system. Progress on the new system is going well, but I don't have an estimate on when it will be ready.

**Comment by icy-arctic-fox** [comment]
The overhauled mock and double system is merged to master. There are some things to cleanup and document, then it will be ready for a release. Until the official docs are available, the [DSL specs](https://github.com/icy-arctic-fox/spectator/tree/master/spec/spectator/dsl/mocks) can be used as a reference.

**Comment by icy-arctic-fox** [comment]
This issue should be resolved in v0.11. There is a [spec](https://github.com/icy-arctic-fox/spectator/blob/master/spec/issues/github_issue_42_spec.cr) dedicated to it. If your issue isn't resolved, please reopen.
